### PR TITLE
Fix issue 18691 - memory errors in std.regex.Captures

### PR DESCRIPTION
**Pull request by DmitryOlshansky**
Instead of a messy code that ended up with refcount at the wrong place
provide a clean sealed fixed-size array with required semantics and
build on top of that.

### CODE REVIEW

File: std/regex/internal/ir.d [hunk 1]
```diff
@@ -899,3 +899,214 @@ struct CharMatcher {
             return trie[ch];
     }
 }
+
+// Internal non-resizeble array, switches between inline storage and CoW
+// POD-only
+struct SmallFixedArray(T, uint SMALL=3)
+if (!hasElaborateDestructor!T)
+{
+    import core.stdc.stdlib : malloc, free;
```
Comment (n8sh) [inline std/regex/internal/ir.d]: How about `core.memory : pureMalloc, pureFree`? Or would there be no benefit because other stuff is impure?

Comment (DmitryOlshansky) [reply std/regex/internal/ir.d]: That’s a good idea. Indeed it won’t make other things pure but worth a shot.

File: std/regex/package.d [hunk 2]
```diff
@@ -681,18 +627,6 @@ public:
 
     ///A hook for compatibility with original std.regex.
     @property ref captures(){ return this; }
-
-    typeof(this) opAssign()(auto ref Captures rhs)
```
Comment (JackStouffer) [inline std/regex/package.d]: don't you want to manually overload this to copy the Store?

Comment (DmitryOlshansky) [reply std/regex/package.d]: Why would I - store is copied + bitblitter by default, no?

**Pull request merged**